source/common/http/conn_manager_impl.cc

```diff
@@ -792,7 +792,8 @@ void ConnectionManagerImpl::ActiveStream::decodeData(ActiveStreamDecoderFilter*
     ENVOY_STREAM_LOG(trace, "decode data called: filter={} status={}", *this,
                      static_cast<const void*>((*entry).get()), static_cast<uint64_t>(status));
     if (!(*entry)->commonHandleAfterDataCallback(status, data, state_.decoder_filters_streaming_)) {
-      return;
+      // break here to ensure that we call decodeTrailers below
```

Comment:
if commonHandleAfterDataCallback fails, doesn't it mean some data filter has requested the pipeline pause iteration?  Let's say we have filter A which adds trailers and filter B which waits for full body and then does an RPC.  If B gets the full body and asks the pipeline to pause while it does a callback, I believe we have to pause (not process trailers) until we get commonContinue.  Are the added trailers not being processed there?

Reply:
Yeah I was a bit unsure how to handle this. The problem I ran into was that the router filter will stop iteration after `decodeData` and keep the connection alive since `end_stream` is false. Since there's no trailers coming in from the decoding codec, nothing ever calls continue, so the the request hangs.
,
Maybe only break when `end_stream` is true to cover this case? In this case we know that we won't get any more data from the codec, so by immediately calling decodeTrailers we're not skipping any data.

Open to other suggestions

Reply:
Actually that might not be feasible since we'd hide some data from filters behind the one that stopped iteration...

Reply:
Oh yeah, I see the problem now.  Hmm.... The best workaround I can think of is fairly inelegant, which is when trailers are added to schedule a closure which basically fakes the "you have received trailers from a client" path but gets called safely from not under the stack of the connection manager pipeline.  Maybe @mattklein123 has a better suggestion?

Reply:
I think this is exactly the same situation that is handled here with body added in the headers callback? https://github.com/envoyproxy/envoy/blob/master/source/common/http/conn_manager_impl.cc#L721. Can we handle in the same way?